src/data/setoid.lean

```diff
@@ -0,0 +1,488 @@
+set.ext $ λ x,
+  ⟨λ hs', symm' (mk_classes c H) $ λ b' hb' h', eq_of_mem_eqv_class H hs hy hb' h' ▸ hs',
+   λ hx, let ⟨b', hc', hb', h'⟩ := H x in
+     (eq_of_mem_eqv_class H hs hy hc' $ hx b' hc' hb').symm ▸ hb'⟩
+
+/-- The equivalence classes of the equivalence relation defined by a set of sets
+    partitioning α are elements of the set of sets. -/
+lemma eqv_class_mem {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b') {y} :
+  {x | rel (mk_classes c H) x y} ∈ c :=
+let ⟨b, hc, hy, hb⟩ := H y in eq_eqv_class_of_mem H hc hy ▸ hc
+
+/-- The union of a set of sets covering α is the entirety of α. -/
+lemma eqv_classes_union {c : set (set α)} (H : ∀ a, ∃ b ∈ c, a ∈ b) :
+  set.sUnion c = @set.univ α :=
+set.univ_subset_iff.1 $ λ x hx, let ⟨b, hm, hb⟩ := H x in set.mem_sUnion_of_mem hb hm
+
+/-- Distinct elements of a set of sets partitioning α are disjoint. -/
+lemma eqv_classes_disjoint {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  (b₁ b₂) (h₁ : b₁ ∈ c) (h₂ : b₂ ∈ c) (h : b₁ ≠ b₂) : disjoint b₁ b₂ :=
+set.disjoint_left.2 $ λ x hx1 hx2, let ⟨b, hc, hx, hb⟩ := H x in
+  h $ eq_of_mem_eqv_class H h₁ hx1 h₂ hx2
+
+/-- A set of disjoint sets covering α partition α (classical). -/
+lemma eqv_classes_of_disjoint_union {c : set (set α)} (hu : set.sUnion c = @set.univ α)
+  (H : ∀ b b' ∈ c, b ≠ b' → disjoint b b') :
+  ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b' :=
+λ a, let ⟨b, hc, ha⟩ := set.mem_sUnion.1 $ show a ∈ _, by rw hu; exact set.mem_univ a in
+  ⟨b, hc, ha, λ b' hc' ha', let hp := classical.prop_decidable in
+  (@not_not _ $ hp _).1 $ λ hn, set.not_mem_empty a $ H b b' hc hc' hn ⟨ha, ha'⟩⟩
+
+/-- If x ∈ α is in 2 elements of a set of disjoint sets covering α, those elements
+    are equal. -/
+lemma eq_of_mem_disjoint {c : set (set α)} (hu : set.sUnion c = @set.univ α)
+  (H : ∀ b b' ∈ c, b ≠ b' → disjoint b b') {x b b'} (hb : b ∈ c) (hx : x ∈ b)
+  (hb' : b' ∈ c) (hx' : x ∈ b') : b = b' :=
+eq_of_mem_eqv_class (eqv_classes_of_disjoint_union hu H) hb hx hb' hx'
+
+/-- Makes an equivalence relation from a set of disjoints sets covering α. -/
+def setoid_of_disjoint_union {c : set (set α)} (hu : set.sUnion c = @set.univ α)
+  (H : ∀ b b' ∈ c, b ≠ b' → disjoint b b') : setoid α :=
+setoid.mk_classes c $ eqv_classes_of_disjoint_union hu H
+
+/-- The equivalence relation made from the equivalence classes of an equivalence
+    relation r equals r. -/
+theorem mk_classes_classes (r : setoid α) :
+  mk_classes r.classes classes_eqv_classes = r :=
+ext' $ λ x y, ⟨λ h, r.symm' (h {z | r.rel z x} (r.mem_classes x) $ r.refl' x),
+  λ h b hb hx, eq_of_mem_classes (r.mem_classes x) (r.refl' x) hb hx ▸ r.symm' h⟩
+
+section partition
+
+def is_partition (α) (c : set (set α)) :=
+∅ ∉ c ∧ ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b'
+
+/-- A partition of α does not contain the empty set. -/
+lemma ne_empty_of_mem_partition {c : subtype (is_partition α)} {s} (h : s ∈ c.1) :
```

Comment:
It's probably better not to use `subtype` here and instead take separate `{c : set (set \a)} (hc : is_partition c)` arguments. As it is now, particularly as `c` is implicit, it would be rather verbose to invoke this lemma if you don't already have `h` of exactly the right form for Lean to infer `c`. Same comment for next two below.

Reply:
I've applied this for the lemmas, and made α implicit since it's obvious from the type of the sets, but for the instances I then have to use ```@is_partition α```. Should α stay explicit? And this change also makes ```partition.le_def``` clunky, but I can't see that anyone would ever want to use it anyway; should I remove it?